### DISCUSSION TITLE
Modern C++

### DISCUSSION THREAD

**Discussion by StefanoD** [CuarzoSoftware/Louvre#18] [General]
Hi,
I am wondering about many raw pointers. I.e. manual pointer destruction in destructors instead of  using `std::unique_ptr`. This can actually cause memory leaks, because the destructor is not called when an exception within the constructor is raised.
Are you aware of that? What are the reasons of using old-school C++?

**Comment by ehopperdietzel** [comment]
Are you pointing out the use of raw pointers to the private implementation of each class? It's indeed a bad design that I haven't addressed, but luckily, it is easy to fix. I was not aware of exceptions, at least I think there is no constructor in the lib that throws or could throw exceptions. As for the second question, I primarily stick to C++11, as I haven't identified a compelling need to adopt newer standards. Is there any particular functionality from the more recent standards that you believe could be beneficial?

**Reply by dblsaiko** [reply]
> there is no constructor in the lib that throws or could throw exceptions

FWIW the `new` operator throws when allocation failed unless you use `new(std::nothrow) type`.

**Reply by ehopperdietzel** [reply]
Is an allocation failure solely due to running out of memory, or are there other circumstances? If it's solely due to running out of memory, it seems like a challenging situation to recover from. Am I mistaken? Additionally, I attempted to replace raw pointers with `std::unique_ptr` for the private implementations of each class. However, `std::unique_ptr` seems to require knowledge of the size of the private structure. This would mean including the private headers in the public headers, which, of course, defeats the purpose entirely. Do you have any ideas on how to navigate this?

**Reply by ehopperdietzel** [reply]
Sorry I was wrong, it is in fact possible to use `std::unique_ptr` https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr/

**Reply by StefanoD** [reply]
I wouldn't recommend you the pimpl-pattern, because it's ugly and has a overhead by having an additional indirection.
Instead, pass a custom deleter function to `std::unique_ptr`: https://stackoverflow.com/questions/19053351/how-do-i-use-a-custom-deleter-with-a-stdunique-ptr-member

**Reply by ehopperdietzel** [reply]
I know that it might scatter memory all over the place, but it doesn't seem to significantly impact performance. I conducted some perf tests, and there aren't many cache misses, at least on my machine. Nevertheless, are there alternative solutions? The private API for each class is extensive and likely to undergo frequent changes in each release. If I make it public, all projects linked to it would require recompilation, which would be quite inconvenient.

**Reply by ehopperdietzel** [reply]
I even considered adding a memory block at the end of each public class with a fixed size and initializing the private class's memory there, but that would indeed be an ugly and risky solution. In any case, public classes mainly generate interfaces, functions, and contain only a pointer to the private class; the entire block of memory with the members is in the private class. Therefore, I believe there would be little difference from including the members in the public class.

**Reply by StefanoD** [reply]
The alternative is, like I said, a custom deleter:

`CustomDeleter.h`:
```cpp
#include <memory>

// Place this in some common header file
template<typename T>
struct Deleter {
    void operator()(T* p) const
    {
        delete p;
    };
};
```

Header of class `B`:
```cpp
#include "CustomDeleter.h"

class A; // Forward declaration

class B {
public:
    B();

private:
    std::unique_ptr<A, Deleter<A>> a;
}
```
cpp of class `B`:
```cpp
#include "B.h"
#include "A.h" // include implementation of A

B::B() : a { std::unique_ptr<A, Deleter<A>>(new A, Deleter<A>()) }
{}
```

Tell me, if this worked. If it didn't work, there is only a small mistake and I can try to create a minimal program.

**Reply by ehopperdietzel** [reply]
Thank you, it's likely to work. However, based on the information provided [here](https://www.fluentcpp.com/2017/09/22/make-pimpl-using-unique_ptr/) I've already succeeded in making it work without the need for a custom deleter. I inserted a print message in the private class destructor, and when the public class goes out of scope, the message is successfully printed.

**Reply by StefanoD** [reply]
So, you have `A` and `AImpl`? And this for every private class? With my solution you don't have this redundancy.

**Reply by ehopperdietzel** [reply]
This is how it would look:

### MyClass.h

```cpp
#include <memory>

class MyClass
{
public:
    MyClass();
    ~MyClass();

    class MyClassPrivate;

    inline MyClassPrivate *imp() const
    {
        return m_imp.get();
    };

private:
    std::unique_ptr<MyClassPrivate> m_imp;
};
```

### MyClass.cpp

```cpp
#include "MyClassPrivate.h"

MyClass::MyClass() : m_imp(new MyClassPrivate)
{
}

MyClass::~MyClass()
{
}
```

**Reply by ehopperdietzel** [reply]
It only requires to define the public class destructor in the .cpp, and the compiler seems to do the magic.

**Reply by StefanoD** [reply]
Ok, I took a closer look into your link and this pimple-unique-ptr-pattern seems like Voodoo. 😅 I understand it, but it is not self-explanatory. Nevertheless, I would maybe also use this pattern in future.

**Reply by ehopperdietzel** [reply]
Great! It's great for encapsulation, but I do wish there were a more elegant solution as well. 😄

**Reply by dblsaiko** [reply]
> MyClass::MyClass() : m_imp(new MyClassPrivate)

Ideally unique_ptr is constructed uisng make_unique:

> MyClass::MyClass() : m_imp(std::make_unique())

Does that work in this position?

**Reply by ehopperdietzel** [reply]
Yes, it works. I read that `std::make_unique` is safer because it performs the allocation atomically. Therefore, I'll use that. I will also need to switch to C++14.

**Reply by StefanoD** [reply]
There are, btw., more generic ways to implement the PIMPL idiom: https://herbsutter.com/gotw/_101/

**Reply by ehopperdietzel** [reply]
Uff that look a bit more complicated, I think the `std::make_unique` approach is simpler and works just fine, so I'll stick with that.

**Comment by StefanoD** [comment]
Yes, I was referring to private implementations and modern C++ starts for me with C++ 11 which also has smart pointers.

I was also wondering about the design choice of virtual classes instead of templates or `std::function` callbacks which are more efficient. But I have to say, that interfaces do have a more beautiful API design. Maybe this is the reason why you chose it.

**Comment by ehopperdietzel** [comment]
That's an excellent question. I had a tough time deciding between the two, including the observer pattern, even though it wouldn't have allowed for a default implementation.

Ultimately, I chose virtual methods because I realized that, while callback functions might be comfortable to work with, they can result in poorly structured projects with things scattered all over the place. Though reimplementing classes and overriding methods can be tedious, it forces you to maintain a well-structured project. Anyone familiar with creating a compositor with Louvre should be able to look at a project and immediately identify where each functionality is handled, or at least get an overall picture of it.

The LAnimation and LTimer classes use callbacks, though.
